freedomofpress/securedrop-client
# Refactor remove controller dependency from dialogs

# Description

This PR builds upon #1374.

#### Current implementation

Currently, the dialog classes send commands to the `Controller` when their buttons are triggered. This results in a two-stages workflow: an _action_ is triggered by the journalist, that action is responsible for ensuring that the user is _authorised_ to perform the action, and to create a confirmation _dialog_  (so far so good). The action depends on the _controller_ because creating the dialog requires it. That dialog, in turn, is responsible for sending the appropriate _controller_ command, which again depends on the controller.

- All the classes involved depend on `Controller`
- The responsibility of performing or not performing the requested action (controller command) is split between the _action_ (when the command should _not_ be sent) and the _dialog_ (when the command is sent).
- Testing the feature is split between the _action_ and the _dialog_ as well.
- Removing the confirmation dialog (UX concern) requires moving the responsibility of sending the controller command from the dialog class to the action.

#### Proposed in this PR

- Actions are responsible for _authorizing_ and _sending_ controller commands. (They hold the dependency on the _controller_)
- Actions can (optionally) create confirmation _dialogs_ for UX reasons. 
- Dialogs classes are responsible for defining the UX flow for the confirmation and ultimately emit the `QDialog` `rejected` and `accepted` signals.
- When a dialog is present, the action relies on those signals as an input into the decision-making before sending the controller command.

As a result:

- Dialogs are purely UI concerns, they don't depend on the _controller_.
- The _action_ is the only class responsible for sending or not sending the controller command.
- The feature can be tested entirely by testing the _action_ class. (Conveniently, the dialog is an injected dependency, so it can be controlled during testing.)
- From the package point of view, the `/actions.py` module depends on the `Controller` class, and most UI classes (widgets) don't. (Eventually, this PR is a first step in that direction.)

# Test Plan

- Confirm that you can delete a **conversation** as usual:
  - [ ] The action "All Files and Messages" is present in the conversation menu 
  - [ ] Triggering the action opens a confirmation dialog
  - [ ] Pressing <kbd>Escape</kbd> aborts the action
  - [ ] Pressing<kbd>Enter</kbd> deletes the conversation
  - [ ] Clicking on the buttons also aborts the action or deletes the conversation as usual

- Confirm that you can delete a **source** as usual:
  - [ ] The action "Entire Source Account" is present in the conversation menu 
  - [ ] Triggering the action opens a confirmation dialog
  - [ ] The confirmation dialog is rendered in "red"
  - [ ] Pressing <kbd>Escape</kbd> aborts the action
  - [ ] Pressing<kbd>Enter</kbd> aborts the action
  - [ ] Clicking on the buttons also aborts the action or deletes the source as usual
  
# Checklist

If these changes modify code paths involving cryptography, the opening of files in VMs or network (via the RPC service) traffic, Qubes testing in the staging environment is required. For fine tuning of the graphical user interface, testing in any environment in Qubes is required. Please check as applicable:


If these changes add or remove files other than client code, the AppArmor profile may need to be updated. Please check as applicable:


If these changes modify the database schema, you should include a database migration. Please check as applicable:
